local Style = script.Parent
local Foundation = script:FindFirstAncestor("foundation")
local Packages = Foundation.Parent

local React = require(Packages.React)
local TagsContext = require(Style.TagsContext)
local getFormattedTags = require(Foundation.Utility.getFormattedTags)

local Types = require(Foundation.Components.Types)
type Tags = Types.Tags

local function useStyleTags(tags: Tags?): string?
	local formattedTags = getFormattedTags(tags)

	local addTags = React.useContext(TagsContext)
	React.useLayoutEffect(function()
		if formattedTags ~= nil then
			addTags(formattedTags)
		end
	end, { formattedTags })
	return formattedTags
end

return useStyleTags
